### PR TITLE
Reader: mishandling of sites that were deleted

### PR DESCRIPTION
**Pull request by samouri**
1. feed-stream had a bad check for errors so was missing cases where site is deleted.
2. get-reader-follows wasn't filtering out follows that had errored.

fixes both:
https://github.com/Automattic/wp-calypso/issues/18139 and https://github.com/Automattic/wp-calypso/issues/18140

### CODE REVIEW

**Comment by samouri** [comment]
To test both fixes:
1.https://calypso.live/read/feeds/44029241?branch=fix/reader/follow-error-feeds
2. Dispatch the action below and ensure it disappears from your manage page 

```js
{
type: 'READER_SITE_REQUEST_FAILURE',
payload: {ID: IdOfSomethingYouFollow}
}
```

File: client/reader/feed-stream/index.jsx [hunk 1]
```diff
@@ -40,7 +40,7 @@ class FeedStream extends React.Component {
 		const emptyContent = <EmptyContent />;
 		const title = getSiteName( { feed, site } ) || this.props.translate( 'Loading Feed' );
 
-		if ( feed && feed.is_error ) {
+		if ( ( feed && feed.is_error ) || ( site && site.is_error ) ) {
```
Comment (samouri) [inline client/reader/feed-stream/index.jsx]: sometimes the feed doesn't error but the site does.  we want to catch both.

File: client/state/reader/sites/reducer.js [hunk 2]
```diff
@@ -46,15 +46,12 @@ function handleDeserialize( state ) {
 
 function handleRequestFailure( state, action ) {
 	// new object proceeds current state to prevent new errors from overwriting existing values
-	return assign(
-		{
-			[ action.payload.ID ]: {
-				ID: action.payload.ID,
-				is_error: true,
-			},
+	return assign( {}, state, {
```
Comment (samouri) [inline client/state/reader/sites/reducer.js]: the way this was written before, `prevState` would override new state.

**Comment by blowery** [comment]
Hmmm, we have a build failure to deal with. 

I think we might want to only overwrite a "good" entry when we get an actual failure from the server (404 only perhaps?). Otherwise, a spotty network could cause us to show sites as broken when they're not and to drop followed sites from the client.

**Comment by samouri** [comment]
1. i'm pretty sure the build errors were in master. I'll rebase and see if its fixed.
2. The way its written now, i don't think we'd overwrite good entries with errors when the network is spotty.  because the payload shouldn't have a list of IDs

**Comment by samouri** [comment]
@jmdodd made a good point that there can be a flicker when going to a site-stream that no longer exists.  I'll make an issue for further improving the ux

**Comment by samouri** [comment]
I've updated this so that it will only remove from the manage/following page in the case of a 403 status code.  Does that make sense @blowery ?

**Comment by blowery** [comment]
> I've updated this so that it will only remove from the manage/following page in the case of a 403 status code.

Should be a 404 (Not Found), not a 403 (Not Authorized).

**Comment by bluefuton** [comment]
> Should be a 404 (Not Found), not a 403 (Not Authorized).

@samouri comments in the code (https://github.com/Automattic/wp-calypso/pull/18142/files#diff-6071751626e9d7aa300e36b6b96afd57R49) that a deleted site will return a 403, rather than the expected 404. Sounds like we should fix the endpoint rather than working around it in the client?

**Comment by fraying** [comment]
@bluefuton @blowery Is this launchable? I just wanna get the follow link off of error pages (#18140). It's important so we don't trigger spam protection.

**Comment by blowery** [comment]
Did some poking. This isn't easily fixable in the API as we handle it at a pretty high level (in `class.wpcom-json-api.php :: process_wpcom_request`. We'll just have to eat the `403` and treat it like a `404`. I think we might force the message to partially match too. So we should validate that it has `'API calls to this endpoint have been disabled` as part of it to turn a 403 into a 404ish.

File: client/reader/feed-stream/index.jsx (hunk 1 above)
Comment (blowery) [reply client/reader/feed-stream/index.jsx]: we may not. there are cases where the feed is ok but the site is in error. For instance, if a site leaves wpcom and the wpcom site is deleted, we'll get an error for the site, but the feed is just fine and should keep working.

**Comment by blowery** [comment]
Had a long conversation with @fraying about this PR and realized we have some confusion as to how sites and feeds should work together. First a little background.

Feeds are the primary thing for the Reader. A feed represents a series of posts, usually from an RSS or Atom feed. A feed is not tied to any sort of hosting and it can change over time. If have a feed from `example.com/feed`, that site could be on wpcom on day 1, external hosting with jetpack on day 2, external hosting without jetpack on day 3 and then back on wpcom on day 4. Feeds don't care.

Sites (or blogs) are a wpcom-ism that augment feeds. A given feed may be associated with a site, and a post in a feed may be associated with a post from a site. We use the wpcom site and post to augment and enhance the data we get from the feed. In addition, many feeds may correspond to the same site.  

With that said, when a site is deleted, we generally _do not_ delete any associated feeds, because many times the delete just means that the feed is no longer hosted by wpcom. Showing a 404 page because a feed's associated site has been deleted is generally the wrong thing to do. There's one subset where it _is_ appropriate (*.wordpress.com sites, which can't move), but generally, we want the feed to stay up. 

This PR is trying to prevent the follow button from appearing on a feed listing, but what we _should_ be doing is changing the API to avoid trying to follow the deleted wpcom site and instead follow the feed by URL.

In addition, we currently prefer site info to feed info when showing various info, like the title. We should disregard the site info entirely if we get back an error and show what we got for the feed. We should also enhance the feed API to return a 404 if the backing site is a *.wordpress.com site and the site has been deleted. We may also want to hook site deletion and remove the feed entirely for such cases, as long as the site has no mapped domains.

This also means we may want a new indicator in the UI that shows that a feed has gone dormant or is failing to update.

**Comment by bluefuton** [comment]
@samouri is this waiting for the '410 Gone' API patch?

**Comment by samouri** [comment]
yessir @bluefuton

**Comment by bluefuton** [comment]
@samouri landed it in wpcom-r163843 👍

File: client/reader/feed-stream/index.jsx (hunk 1 above)
Comment (samouri) [reply client/reader/feed-stream/index.jsx]: hmm. then how do we "disconnect" the site/feed?

File: client/reader/feed-stream/index.jsx [hunk 3]
```diff
@@ -41,7 +41,7 @@ class FeedStream extends React.Component {
 		const emptyContent = <EmptyContent />;
 		const title = getSiteName( { feed, site } ) || this.props.translate( 'Loading Feed' );
 
-		if ( feed && feed.is_error ) {
+		if ( ( feed && feed.is_error ) || ( site && site.is_error && site.error.code === 410 ) ) {
```
Comment (blowery) [inline client/reader/feed-stream/index.jsx]: good. i think we just want to drop the row now.

Comment (blowery) [reply client/reader/feed-stream/index.jsx]: nevermind, this is the feed stream, not the manage page.

**Comment by bluefuton** [comment]
✅ Removed site from Manage Following on receiving a 410
✅ Received a proper error page when receiving a 410 on feed stream:

screen shot 2017-10-24 at 11 08 21

**Pull request merged**